File: app/src/main/java/org/thoughtcrime/securesms/stickers/StickerKeyboardPageViewModel.java
package org.thoughtcrime.securesms.stickers;

import android.app.Application;
import android.database.ContentObserver;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import org.thoughtcrime.securesms.database.DatabaseContentProviders;
import org.thoughtcrime.securesms.database.model.StickerRecord;
import org.thoughtcrime.securesms.util.Throttler;

import java.util.List;

final class StickerKeyboardPageViewModel extends ViewModel {

  static final String RECENT_PACK_ID = "RECENT";

  private final Application                          application;
  private final StickerKeyboardRepository            repository;
  private final MutableLiveData<List<StickerRecord>> stickers;
  private final Throttler                            observerThrottler;
  private final ContentObserver                      observer;

  private String packId;

  private StickerKeyboardPageViewModel(@NonNull Application application, @NonNull StickerKeyboardRepository repository) {
    this.application       = application;
    this.repository        = repository;
    this.stickers          = new MutableLiveData<>();
    this.observerThrottler = new Throttler(500);
    this.observer          = new ContentObserver(new Handler(Looper.getMainLooper())) {
      @Override
      public void onChange(boolean selfChange) {
        observerThrottler.publish(() -> getStickers(packId));
      }
    };

    application.getContentResolver().registerContentObserver(DatabaseContentProviders.Sticker.CONTENT_URI, true, observer);
  }

  LiveData<List<StickerRecord>> getStickers(@NonNull String packId) {
    this.packId = packId;

    if (RECENT_PACK_ID.equals(packId)) {
      repository.getRecentStickers(stickers::postValue);
    } else {
      repository.getStickersForPack(packId, stickers::postValue);
    }

    return stickers;
  }

  @Override
  protected void onCleared() {
    application.getContentResolver().unregisterContentObserver(observer);
  }

  static class Factory extends ViewModelProvider.NewInstanceFactory {
    private final Application               application;
    private final StickerKeyboardRepository repository;

    Factory(@NonNull Application application, @NonNull StickerKeyboardRepository repository) {
      this.application = application;
      this.repository  = repository;
    }

    @Override
    public @NonNull<T extends ViewModel> T create(@NonNull Class<T> modelClass) {
      //noinspection ConstantConditions
      return modelClass.cast(new StickerKeyboardPageViewModel(application, repository));
    }
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/keyboard/sticker/StickerKeyboardPageViewModel.kt
package org.thoughtcrime.securesms.keyboard.sticker

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import org.thoughtcrime.securesms.database.SignalDatabase
import org.thoughtcrime.securesms.keyboard.sticker.KeyboardStickerPackListAdapter.StickerPack
import org.thoughtcrime.securesms.util.MappingModelList
import org.thoughtcrime.securesms.util.livedata.LiveDataUtil

private const val NO_SELECTED_PAGE = "no_selected_page"

class StickerKeyboardPageViewModel(private val repository: StickerKeyboardRepository) : ViewModel() {
  private val keyboardStickerPacks: MutableLiveData<List<StickerKeyboardRepository.KeyboardStickerPack>> = MutableLiveData()

  private val selectedPack: MutableLiveData<String> = MutableLiveData(NO_SELECTED_PAGE)

  val packs: LiveData<List<StickerPack>>
  val stickers: LiveData<MappingModelList>

  init {
    val stickerPacks: LiveData<List<StickerPack>> = LiveDataUtil.mapAsync(keyboardStickerPacks) { packs ->
      packs.map { StickerPack(it) }
    }

    packs = LiveDataUtil.combineLatest(selectedPack, stickerPacks) { selected, packs ->
      if (packs.isEmpty()) {
        packs
      } else {
        val actualSelected = if (selected == NO_SELECTED_PAGE) packs[0].packRecord.packId else selected
        packs.map { it.copy(selected = it.packRecord.packId == actualSelected) }
      }
    }

    stickers = LiveDataUtil.mapAsync(keyboardStickerPacks) { packs ->
      val list = MappingModelList()

      packs.forEach { pack ->
        list += KeyboardStickerListAdapter.StickerHeader(pack.packId, pack.title, pack.titleResource)
        list += pack.stickers.map { KeyboardStickerListAdapter.Sticker(pack.packId, it) }
      }

      list
    }
  }

  fun getSelectedPack(): LiveData<String> = selectedPack

  fun selectPack(packId: String) {
    selectedPack.value = packId
  }

  fun refreshStickers() {
    repository.getStickerPacks { keyboardStickerPacks.postValue(it) }
  }

  class Factory(context: Context) : ViewModelProvider.Factory {
    private val repository = StickerKeyboardRepository(SignalDatabase.stickers)

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
      return requireNotNull(modelClass.cast(StickerKeyboardPageViewModel(repository)))
    }
  }
}
